mmdet/models/detectors/base_detr.py

```diff
@@ -0,0 +1,224 @@
+            **head_inputs_dict,
+            rescale=rescale,
+            batch_data_samples=batch_data_samples)
+        batch_data_samples = self.add_pred_to_datasample(
+            batch_data_samples, results_list)
+        return batch_data_samples
+
+    def _forward(
+            self,
+            batch_inputs: Tensor,
+            batch_data_samples: OptSampleList = None) -> Tuple[List[Tensor]]:
+        """Network forward process. Usually includes backbone, neck and head
+        forward without any post-processing.
+
+         Args:
+            batch_inputs (Tensor): Inputs with shape (N, C, H, W).
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`.
+                Defaults to None.
+
+        Returns:
+            tuple[Tensor]: A tuple of features from ``bbox_head`` forward.
+        """
+        img_feats = self.extract_feat(batch_inputs)
+        head_inputs_dict = self.forward_transformer(img_feats,
+                                                    batch_data_samples)
+        results = self.bbox_head.forward(
+            **head_inputs_dict)  # TODO: refine this  # noqa
+        return results
+
+    def forward_transformer(self, img_feats,
+                            batch_data_samples) -> Dict:  # TODO: typehint
+        # TODO: Doc
+        encoder_inputs_dict, decoder_inputs_dict = self.pre_transformer(
+            img_feats, batch_data_samples)
+
+        encoder_outputs_dict = self.forward_encoder(**encoder_inputs_dict)
+
+        tmp_dec_in, head_inputs_dict = self.pre_decoder(**encoder_outputs_dict)
+        decoder_inputs_dict.update(tmp_dec_in)  # TODO: refine 'update'
+
+        decoder_outputs_dict = self.forward_decoder(**decoder_inputs_dict)
+        head_inputs_dict.update(decoder_outputs_dict)
+        return head_inputs_dict
+
+    def extract_feat(self, batch_inputs: Tensor) -> Tuple[Tensor]:
+        """Extract features.
+
+        Args:
+            batch_inputs (Tensor): Image tensor with shape (N, C, H ,W).
+
+        Returns:
+            tuple[Tensor]: Multi-level features that may have
+                different resolutions.
+        """
+        x = self.backbone(batch_inputs)
+        if self.with_neck:
+            x = self.neck(x)
+        return x
+
+    @abstractmethod
+    def pre_transformer(
+        self,
+        img_feats: Tuple[Tensor],
+        batch_data_samples: OptSampleList = None
+    ) -> Tuple[Dict, Dict]:  # TODO: typehint
+        """This function creates the inputs of the Transformer.
+
+        1. Construct batch padding mask.
+        2. Prepare transformer_inputs_dict.
+        """
+        pass
+
+    @abstractmethod
+    def forward_encoder(self, **kwargs) -> Tensor:  # TODO: typehint
+        """TODO: Doc"""
+        pass
+
+    @abstractmethod
+    def pre_decoder(self,
```

Comment:
Seems we should pass `batch_data_samples`？ Such as dino will use it to generate noisy queries

Reply:
We can overload `self.forward_transformer()` in `DINO` and modified this line as:
```Python3
tmp_dec_in, head_inputs_dict = self.pre_decoder(**encoder_outputs_dict, 
                                                batch_data_samples=batch_data_samples)
```

or prepare `dn_query_bboxes` and `dn_query_labels` in the `self.pre_transformer()` and store them in `decoder_inputs_dict`.

There is no need for `batch_data_samples` in a non-DN DETR model, I prefer to make custom modifications in DN DETRs.